python/ray/serve/_private/proxy.py

```diff
@@ -374,144 +445,76 @@ async def proxy_request(self, proxy_request: ProxyRequest) -> ProxyResponse:
         """
         assert proxy_request.request_type in {"http", "websocket", "grpc"}
 
-        method = proxy_request.method
-
-        # only use the non-root part of the path for routing
-        route_path = proxy_request.route_path
-
-        if proxy_request.is_route_request:
-            if self._is_draining():
-                return await self.draining_response(proxy_request=proxy_request)
-
-            self.request_counter.inc(
-                tags={
-                    "route": route_path,
-                    "method": method,
-                    "application": "",
-                    "status_code": self.success_status_code,
-                }
-            )
-            return await self.routes_response(proxy_request=proxy_request)
-
-        if proxy_request.is_health_request:
-            if self._is_draining():
-                return await self.draining_response(proxy_request=proxy_request)
+        response_handler_info = self._get_response_handler_info(proxy_request)
 
-            self.request_counter.inc(
-                tags={
-                    "route": route_path,
-                    "method": method,
-                    "application": "",
-                    "status_code": self.success_status_code,
-                }
-            )
-            return await self.health_response(proxy_request=proxy_request)
+        start_time = time.time()
+        if response_handler_info.should_increment_ongoing_requests:
+            self._ongoing_requests_start()
 
         try:
-            self._ongoing_requests_start()
+            # The final message yielded must always be the `ResponseStatus`.
```

Comment:
The code doesn't check the final message is always `ResponseStatus`. i would suggest raise internal error if the final message is required.

Reply:
Assertion is at line 463 😄

Reply:
but it is not guaranteed to be the final message?

Reply:
Oh I guess I can add a break after status is set. Let me do that real quick

Reply:
why do you need a break?

Reply:
To ensure that the last message is `ResponseStatus` type. Functionally should be the same.

Reply:
But will leave this out until we reach a consensus